[Docs] How to use with PyTorch page (#4474)
huggingface/datasets: docs/source/use_with_pytorch.mdx

# Use with PyTorch

> ## Tensors

Reviewer: I think we can remove the `tensors` header since it doesn't have any associated content under it. It would also be nice to write a brief intro under `# Use with PyTorch` about what this doc covers and how it will help users.

### Dataset format

By default, datasets return regular python objects: integers, floats, strings, lists, etc.

> To get PyTorch tensors instead, you can set the format of the dataset to "pytorch" using [`Dataset.with_format`]:

Reviewer: Small discrepancy with the `with_format` docstring which says it accepts `torch` instead of `pytorch`. Should we update the docstring to align this?

Author: Both are supported, but let me change this so that we use the same everywhere for consistency

> To get PyTorch tensors instead, you can set the format of the dataset to "pytorch" using [`Dataset.with_format`]:

Reviewer: ```suggestion
To get PyTorch tensors instead, you can set the format of the dataset to `pytorch` using [`Dataset.with_format`]:
```

```py
>>> from datasets import Dataset
>>> data = [[1, 2],[3, 4]]
>>> ds = Dataset.from_dict({"data": data})
>>> ds = ds.with_format("pytorch")
>>> ds[0]
{'data': tensor([1, 2])}
>>> ds[:2]
{'data': tensor([[1, 2],
         [3, 4]])}
```

<Tip>

A [`Dataset`] object is a wrapper of an Arrow table, which allows fast zero-copy reads from arrays in the dataset to PyTorch tensors.

</Tip>

> To load the data as tensors on a GPU, you can also specify the `device` argument:

Reviewer: ```suggestion
To load the data as tensors on a GPU, specify the `device` argument:
```

```py
>>> import torch
>>> device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
>>> ds = ds.with_format("pytorch", device=device)
>>> ds[0]
{'data': tensor([1, 2], device='cuda:0')}
```

### N-dimensional arrays

If your dataset consists of N-dimensional arrays, you will see that by default they are considered as nested lists.

> In particular, a PyTorch formatted dataset outputs nested lists instead of one single tensor:

Reviewer: ```suggestion
In particular, a PyTorch formatted dataset outputs nested lists instead of a single tensor:
```

```py
>>> from datasets import Dataset
>>> data = [[[1, 2],[3, 4]],[[5, 6],[7, 8]]]
>>> ds = Dataset.from_dict({"data": data})
>>> ds = ds.with_format("pytorch")
>>> ds[0]
{'data': [tensor([1, 2]), tensor([3, 4])]}
```

> To get one single tensor, you must explicitly use the Array feature type and specify the shape of your tensors:

Reviewer: ```suggestion
To get a single tensor, you must explicitly use the [`Array`] feature type and specify the shape of your tensors:
```

>>> ds[:2]                                                                                                                                                     
{'data': tensor([0, 1]), 'text': ['foo', 'bar']}
```

The [`Image`] and [`Audio`] feature types are not supported yet.

## Data loading

Like `torch.utils.data.Dataset` objects, a [`Dataset`] can be passed directly to a PyTorch `DataLoader`:

```py
>>> import numpy as np
>>> from datasets import Dataset 
>>> from torch.utils.data import DataLoader
>>> data = np.random.rand(16)
>>> label = np.random.randint(0, 2, size=16)
>>> ds = Dataset.from_dict({"data": data, "label": label}).with_format("pytorch")
>>> dataloader = DataLoader(ds, batch_size=4)
>>> for batch in dataloader:
...     print(batch)                                                                                            
{'data': tensor([0.0047, 0.4979, 0.6726, 0.8105]), 'label': tensor([0, 1, 0, 1])}
{'data': tensor([0.4832, 0.2723, 0.4259, 0.2224]), 'label': tensor([0, 0, 0, 0])}
{'data': tensor([0.5837, 0.3444, 0.4658, 0.6417]), 'label': tensor([0, 1, 0, 0])}
{'data': tensor([0.7022, 0.1225, 0.7228, 0.8259]), 'label': tensor([1, 1, 1, 1])}
```

> ### Optimize data loading

Reviewer: Feel free to write a brief intro about optimizing data loading, otherwise, I would remove this section header since it doesn't have any content associated with it. Maybe something along the lines of:

There are several ways you can increase the speed your data is loaded which can save you time, especially if you are working with large datasets. PyTorch offers parallelized data loading, retrieving batches of indices instead of individually, and streaming to progressively download datasets.

> #### Use multiple Workers

Reviewer: ```suggestion
#### Use multiple workers
```

> You can parallelize data loading the `num_workers` argument of a PyTorch `DataLoader` and get a higher throughput.

Reviewer: ```suggestion
You can parallelize data loading with the `num_workers` argument of a PyTorch `DataLoader` and get a higher throughput.
```

Under the hood, the `DataLoader` starts `num_workers` processes.
Each process reloads the dataset passed to the `DataLoader` and is used to query examples.

> Reloading the dataset inside a worker doesn't fill up your RAM, since it simply memory-map the dataset again from your disk.

Reviewer: ```suggestion
Reloading the dataset inside a worker doesn't fill up your RAM, since it simply memory-maps the dataset again from your disk.
```

```py
>>> import numpy as np
>>> from datasets import Dataset, load_from_disk
>>> from torch.utils.data import DataLoader
>>> data = np.random.rand(10_000)
>>> Dataset.from_dict({"data": data}).save_to_disk("my_dataset")
>>> ds = load_from_disk("my_dataset").with_format("torch")
>>> dataloader = DataLoader(ds, batch_size=32, num_workers=4)
```

#### Use a BatchSampler

> By default the pytorch data loader load batches of data from a dataset one by one like this:

Reviewer: ```suggestion
By default, the PyTorch `DataLoader` load batches of data from a dataset one by one like this:
```

```py
batch = [dataset[idx] for idx in range(start, end)]
```

Unfortunately, this does numerous read operations on the dataset.
It is more efficient to query batches of examples using a list:

```py
batch = dataset[start:end]
# or
batch = dataset[list_of_indices]
```

> For the pytorch data loader to query batches using a list, you can use a `BatchSampler`:

Reviewer: ```suggestion
For the PyTorch `DataLoader` to query batches using a list, you can use a `BatchSampler`:
```

```py
>>> from torch.utils.data.sampler import BatchSampler, RandomSampler
>>> sampler = BatchSampler(RandomSampler(ds), batch_size=32, drop_last=False)
>>> dataloader = DataLoader(ds, sampler=sampler)
```

Moreover, this is particularly useful if you used [`set_transform`] to apply a transform on-the-fly when examples are accessed.
You must use a `BatchSampler` if you want the transform to be given full batches instead of receiving `batch_size` times one single element.

### Stream data

Loading a dataset in streaming mode is useful to progressively download the data you need while iterating over the dataset.

> Set the format of a streaming dataset to "pytorch", and it inherits from `torch.utils.data.IterableDataset` so you can pass it to a DataLoader:

Reviewer: ```suggestion
Set the format of a streaming dataset to `pytorch`, and it inherits from `torch.utils.data.IterableDataset` so you can pass it to a `DataLoader`:
```
